Some refactoring

ichthyosaurus opened the pull request:
- all-new app icon (color theme is inspired by the original Wunderfitz icon)
- cover page: refactored layout, removed background
- small changes in the backend to use properties instead of invokables
- new info menu item telling the user that the Heinzelnisse dict cannot be deleted
- new tab bar implementation in a separate module (can be reused)
- some minor refactoring in the two main views
- some cleanup in the curiosity view
- show keyboard immediately after statup
- new settings option to disable cloud features
- updated German translation
- updated other translation catalogs (no new translations)

I split the TitlePage.qml component into multiple files (TitlePage.qml, DictionariesView.qml, CuriosityView.qml). That makes it harder to see my changes to the views, sorry :(.

This PR makes my fork obsolete.

Wunderfitz commented:
Wow, that's really a lot for "some refactoring". Thank you very much, really appreciated! :) 

On the other side, I'm terribly sorry, but it's nearly impossible to properly review a pull request that addresses so many different areas and also mixes several things. I would integrate some things really fast, others I'd like to discuss and for some I wonder if they are really needed.

Please split your PR and post separate PRs per topic.

Some hints where you need good persuasion skills to get them merged:
- Script to generate images. Though it's impressive to have such a script, but as they rarely change, why is it actually needed? Why not simply include the files in the repo and also please don't remove the 256px version...
- Removal of the cover background & additional info "Wunderfitz Dictionary". The new list header is cool, but why the rest? I'd prefer to keep it clean... 
- Why two different settings for cloud-enablement and cloud-terms accepted?
- Disabled menu item saying that the Heinzelnisse dict can't be deleted. That better should go somewhere else or kept like it it

In general as also already mentioned previously: I prefer to keep it simple (and stupid). Some elements seem to make it quite complicated (e.g. in the new docked tab bar) and add more coding than there was before - probably hard to debug and also to understand... And one thing as well for the tab bar: If you license it that way in the file header, you need to make it available as a separate component somewhere. Otherwise, you need to write "Wunderfitz" there as parent project as well.

ichthyosaurus commented:
> Please split your PR and post separate PRs per topic.

I tried to do that several times but it is almost impossible. There are major changes to the structure of modules and the changes are interdependent.

> but it's nearly impossible to properly review a pull request that addresses so many different areas and also mixes several things.

Yes, I know that. Sadly, that lies a bit in the nature of refactoring... I already tried cherry-picking commits. You can give it a try too with commits you like, but you will see that's also not quite possible because of the bigger changes.

My commit messages are quite explanatory, though. I would recommend roughly following the commits to review the changes and see what went where. It's only ~40 commits, so reviewing should be manageable this way.

> I'd like to discuss and for some I wonder if they are really needed.

Let's discuss them here and revert the relevant commits/changes if necessary.

To the concerns you already mentioned:

> Script to generate images. Though it's impressive to have such a script, but as they rarely change, why is it actually needed?

Such a script is immensely helpful for iterative development of icons/graphics. Sure, it is less important to have it once an icon is finished, but deleting it would mean having to do the work again if one decides to change the graphics later. It automates the tedious and repetitive tasks of exporting all changed files, and shrinking the generated files.

> Why not simply include the files in the repo

They are included, of course...?

> and also please don't remove the 256px version...

I added it again. It's a non-standard size on Sailfish, so I removed it before. Why do you need it?

> Removal of the cover background & additional info "Wunderfitz Dictionary". The new list header is cool, but why the rest? I'd prefer to keep it clean...

I tried several options and decided this was the most clean and uncluttered way.
- having the background visible all the time made the results hardly legible
- showing the background only when no results are shown felt confusing and cluttered
- showing an empty list when no results are shown looked too empty

To my aesthetic perception this was the most appealing way.

> Why two different settings for cloud-enablement and cloud-terms accepted?

Because the tab bar is hidden when cloud features are disabled, so the cloud view cannot be accessed. After enabling the cloud integration, the user still has to explicitly accept the terms and conditions before being able to use it. The request that's already being shown has to be reflected in the settings as well.

> Disabled menu item saying that the Heinzelnisse dict can't be deleted. That better should go somewhere else or kept like it it

Yes and no. Sailfish's style guide speaks against showing disabled pulley entries. When a user-installed dictionary is selected, the user expects to be able to delete it via the pulley menu. Now let's say the user switches to a built-in dictionary and wants to delete it (e.g. to save space). How we could handle it:

- disable the pulley so it is visible but can't be accessed: visually misleading and confusing → no
- remove the pulley to avoid confusion: looks like a visual glitch when the pulley hint suddenly disappears → no
- show an explanatory menu entry: visually the same and in a place where the user expects it, but against the style guide → maybe

I would always recommend following the style guide when we're talking about a menu with multiple entries. In this particular case (menu with one entry, changes upon changed selection in a different menu) it still appears as the best option to me to disable the entry.

You could simply revert commit f32eb06, though.

> I prefer to keep it simple (and stupid).

Yes, though the KISS principle can be followed in different ways. To me, less code repetition means less code, meaning easier maintenance and better clarity - so to me, re-using abstract components means following KISS. 

> Some elements seem to make it quite complicated (e.g. in the new docked tab bar) and add more coding than there was before - probably hard to debug and also to understand...

About the tab bar: it is much more flexible (and less buggy) than before, most importantly: it can be reused in other applications. Of course this added flexibility comes at the cost of a few LOC. I'm still convinced that having a separate, reusable component is better than having the implementation scattered over one giant QML file.

What other elements do you mean? Can you be more specific?

> And one thing as well for the tab bar: If you license it that way in the file header, you need to make it available as a separate component somewhere. Otherwise, you need to write "Wunderfitz" there as parent project as well.

Of course; it is available here: https://github.com/ichthyosaurus/sf-docked-tab-bar


So I'd like to ask you to re-open this PR and review the changes following the line of commits. If you are a bit more specific, I'd like to address your remaining concerns.

ichthyosaurus commented:
I re-added the 256px icon but the commit won't show up here because the PR is closed.

Wunderfitz commented:
Hey there, long time no see... ;)

I guess I'd need to check everything once again to be able to answer properly. Of course, I reopened the PR, but I also have to say that development of the dictionary app is currently not that high on my priority list. I'm fully busy with my Telegram app and also want/need to focus on it. 

To be honest: I currently don't see how I could justify such a massive change to my spare time budget. That might change in the future, but at the moment the cost/benefit ratio doesn't fit. 

Also in the long run: In the end I need to maintain the change if I merge it - and for that it's vital that I understand the code or am at least able to debug/reproduce the steps there. If I recall my review in summer correctly, that might be very tricky. Anyway, that's the current state of the art. I'm sorry that I don't have another answer for you.

Best regards, Sebastian

ichthyosaurus commented:
Yep, I see your point :). Of course I would like to see it merged, first because I already strained my spare time budget with this, and second because I'd like to get rid of my fork ;).
